hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/INodeWithAdditionalFields.java

```diff
@@ -486,4 +498,239 @@ private int getGroupIDDB(String name) throws IOException {
+        } break;
+        case PROJECT:
+        case OTHER:
+        default: return;
+      }
+    }
+  }
+  
+  private void logProvenanceEvent(TrackedProv type, INodeDirectory[] parents, FileProvenanceEntry.Operation op,
+    Optional<XAttr> xattr) throws TransactionContextException, StorageException, IOException {
+    
+    Long projectIId;
+    String projectName;
+    switch(type) {
+      case DATASET: {
+        projectIId = parents[ProvParents.PROJECT.ordinal()].getId();
+        projectName = parents[ProvParents.PROJECT.ordinal()].getLocalName();
+      } break;
+      case HIVE: {
+        projectIId = -1l;
+        projectName = ProvenanceHelper.projectNameFromHiveDB(
+          parents[ProvParents.DATASET.ordinal()].getLocalName());
+      } break;
+      default: return;
+    }
+    
+    UserGroupInformation ugi;
+    int remoteUserId;
+    String remoteUserName;
+    try {
+      ugi = NameNode.getRemoteUser();
+      remoteUserId = UsersGroups.getUserID(ugi.getUserName());
+      remoteUserName = ugi.getUserName();
+    } catch (IOException ex) {
+      String msg = "provenance - error getting user - issuer of operation for inode:" + this.getLocalName();
+      LOG.error(msg, ex);
+      throw ex;
+    }
+    String appId = ugi.getApplicationId();
+    if(appId == null) {
+      appId = "none";
+    }
+    
+    long timestamp = System.currentTimeMillis();
+    String p1Name = parents[ProvParents.PARENT_P1.ordinal()] != null
+      ? parents[ProvParents.PARENT_P1.ordinal()].getLocalName() : "";
+    String p2Name = parents[ProvParents.PARENT_P2.ordinal()] != null
+      ? parents[ProvParents.PARENT_P2.ordinal()].getLocalName() : "";
+    
+    String xattrName = xattr.isPresent() ? xattr.get().getName() : "";
+    
+    FileProvenanceEntry ple = new FileProvenanceEntry(id, op, logicalTime, timestamp, appId, remoteUserId,
+      partitionId, projectIId, parents[ProvParents.DATASET.ordinal()].getId(),
+      parents[ProvParents.PARENT_DIRECT.ordinal()].getId(),
+      getLocalName(), projectName, parents[ProvParents.DATASET.ordinal()].getLocalName(),
+      p1Name, p2Name, parents[ProvParents.PARENT_DIRECT.ordinal()].getLocalName(),
+      remoteUserName, xattrName, logicalTime, timestamp, parents[ProvParents.DATASET.ordinal()].getLogicalTime());
+    
+    try {
+      if (xattr.isPresent()) {
+        byte xattrNamespace = xattr.get().getNameSpace().getId();
+        byte[] xattrValue = xattr.get().getValue();
+        FileProvXAttrBufferEntry xattrEntry
+          = new FileProvXAttrBufferEntry(id, xattrNamespace, xattrName, logicalTime, xattrValue);
+        EntityManager.add(xattrEntry);
+      }
+      EntityManager.add(ple);
+    } catch (TransactionContextException | StorageException ex) {
+      String msg = "provenance - error persisting in ndb for inode:" + this.getLocalName();
+      LOG.error(msg, ex);
+      throw ex;
+    }
+  }
+  
+  private static class ProvenanceHelper {
```

Comment:
wouldn't it be better to put it in its own file/class?

Reply:
I second that

Reply:
done.
moved everything to FileProvenance class and you know call FileProvenance.log(inode, ...)

Reply:
added proxy methods in the INodeWithAdditionalFields class, but they are just that - proxy fields to the FileProvenance.log methods.